Sources/NIOCore/EventLoop.swift

```diff
@@ -356,6 +356,14 @@ public protocol EventLoop: EventLoopGroup {
     /// It is valid for an `EventLoop` not to implement any of the two `_promise` functions. If either of them are implemented,
     /// however, both of them should be implemented.
     func _promiseCompleted(futureIdentifier: _NIOEventLoopFutureIdentifier) -> (file: StaticString, line: UInt)?
+
+    /// Set a timer that will call a handler at the given time.
+    @discardableResult
+    func setTimer(for deadline: NIODeadline, _ handler: any NIOTimerHandler) -> NIOTimer
```

Comment:
Starting a separate thread to bikeshed how we name or document the requirements of this API.

Specifically, this API only tolerates being called on the event loop. Failure to do so should crash (I need to add the precondition in).

We discussed briefly off-PR naming these `unsafe*` but we concluded that this is a misuse of the "unsafe" terminology, which is reserved for undefined behaviour. Furthermore, these APIs are very safe because they will crash if used incorrectly. This is more like checked vs. unchecked (like with continuations and arithmetic) but we don't really want to provide an unchecked variant.

So we have some options (and I'm open to others). In all cases we will document the requirements:
1. Keep `setTimer(for:_:)` and rely only on documentation for the caller requirements.
2. `setCheckedTimer(for:_:)`.
3. `setTimerSync(for:_:)`.
4. `setTimer(for:_:iAmCallingThisFromTheEventLoop: Bool)`

My thoughts:

(1) seems in keeping with other Swift APIs that do runtime checks (e.g. arithmetic).
(2) would be in keeping with other Swift APIs that do runtime checks but only make sense if we were providing an unchecked counterpart, which we don't want to do.
(3) There's _some_ (not much) president for this "sync" terminology in NIO API but it predates `async` and so is arguably confusing now. This goes double for a timer API and could be misunderstood as a call that blocks until the timer fires.
(4) This just creates noise when used properly and still fatal errors when used incorrectly. So lose-lose.

My current conclusion is to keep the name as-is, but I'm happy to discuss it further.

Reply:
OK, so that means we don't concern ourselves with any augmented API symbols to communicate the contract. I'll add some documentation to `NIOCustomTimerImplementation` methods that implementations _should_ tolerate being called from off the EL.